### PR TITLE
Transformation rule nodes enhancements

### PR DESCRIPTION
**Pull request by ShvaykaD**
## Pull Request description

 - Transformation nodes enhancements(docs + minor refactoring + added tests for changed logic).
 - Improved error handling on init for fetchTo based nodes
 - see examples templatization docs. Files created with common content only. No examples added.

PE PR: https://github.com/thingsboard/thingsboard-pe/pull/2054

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [x] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [x] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [x] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java [hunk 1]
```diff
@@ -21,89 +21,101 @@
...
         configClazz = TbCopyKeysNodeConfiguration.class,
-        nodeDescription = "Copies the msg or metadata keys with specified key names selected in the list",
-        nodeDetails = "Will fetch fields values specified in list. If specified field is not part of msg or metadata fields it will be ignored." +
-                "Returns transformed messages via <code>Success</code> chain",
+        nodeDescription = "Copies key-values from message to message metadata or vice-versa.",
+        nodeDetails = "Fetches key-values from message or message metadata based on the keys list specified in the configuration " +
+                "and copies them into message metadata or into message itself in accordance with the fetch source. " +
+                "Keys that are absent in the fetch source will be ignored. " +
+                "Use regular expression(s) as a key(s) to copy keys by pattern.<br><br>" +
+                "Output connections: <code>Success</code>, <code>Failure</code>.",
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java:50]: Copies specified key-value pairs from the message body to metadata, or vice-versa, according to the configured direction and keys. Regular expressions can be used to define which keys-value pairs to copy. Any configured key not found in the source will be ignored.

Output connections:
Success - The node successfully processed and forwarded the message.
Failure - The node encountered an issue processing the message.

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java [hunk 2]
```diff
@@ -21,89 +21,101 @@
...
         uiResources = {"static/rulenode/rulenode-core-config.js"},
         configDirective = "tbTransformationNodeCopyKeysConfig",
         icon = "content_copy"
 )
-public class TbCopyKeysNode implements TbNode {
+public class TbCopyKeysNode extends TbAbstractTransformNodeWithTbMsgSource {
 
     private TbCopyKeysNodeConfiguration config;
     private List<Pattern> patternKeys;
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java:58]:
```suggestion
    private List<Pattern> compiledKeyPatterns;
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java [hunk 3]
```diff
@@ -21,89 +21,101 @@
...
-                    String keyData = entry.getKey();
-                    if (checkKey(keyData)) {
-                        msgChanged = true;
-                        msgDataNode.put(keyData, entry.getValue());
+            switch (copyFrom) {
+                case METADATA:
+                    ObjectNode msgDataNode = (ObjectNode) dataNode;
+                    Map<String, String> metaDataMap = metaDataCopy.getData();
+                    for (Map.Entry<String, String> entry : metaDataMap.entrySet()) {
+                        String keyData = entry.getKey();
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java:83]:
```suggestion
                        String key = entry.getKey();
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java [hunk 4]
```diff
@@ -21,89 +21,101 @@
...
-                        String value = entry.getValue().isTextual() ?
-                                entry.getValue().asText() : JacksonUtil.toString(entry.getValue());
-                        metaData.putValue(keyData, value);
+                    msgData = JacksonUtil.toString(msgDataNode);
+                    break;
+                case DATA:
+                    Iterator<Map.Entry<String, JsonNode>> iteratorNode = dataNode.fields();
+                    while (iteratorNode.hasNext()) {
+                        Map.Entry<String, JsonNode> entry = iteratorNode.next();
+                        String keyData = entry.getKey();
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbCopyKeysNode.java:95]:
```suggestion
                        String key = entry.getKey();
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java [hunk 5]
```diff
@@ -21,84 +21,96 @@
...
+                            keysToDelete.add(keyData);
+                        }
+                    });
+                    msgDataObject.remove(keysToDelete);
+                    msgData = JacksonUtil.toString(msgDataObject);
+                }
+                break;
+            default:
+                log.debug("Unexpected DeleteFrom value: {}. Allowed values: {}", deleteFrom, TbMsgSource.values());
+                break;
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java:102]:
```suggestion
```
I think break is redundant as a last statement in a default case

Comment (dskarzh) [reply rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java:102]: Please check other nodes

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java [hunk 6]
```diff
@@ -21,84 +21,96 @@
...
+        name = "delete key-values",
+        version = 1,
         configClazz = TbDeleteKeysNodeConfiguration.class,
-        nodeDescription = "Removes keys from the msg data or metadata with the specified key names selected in the list",
-        nodeDetails = "Will fetch fields (regex) values specified in list. If specified field (regex) is not part of msg " +
-                "or metadata fields it will be ignored. Returns transformed messages via <code>Success</code> chain",
+        nodeDescription = "Removes key-values from message or message metadata.",
+        nodeDetails = "Removes key-values from message or message metadata based on the keys list specified in the configuration. " +
+                "Use regular expression(s) as a key(s) to remove keys by pattern.<br><br>" +
+                "Output connections: <code>Success</code>, <code>Failure</code>.",
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java:49]: This node deletes specified key-value pairs from the message body or metadata. Keys and/or regular expressions specified in configuration are used to determine which keys should be removed.

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java [hunk 7]
```diff
@@ -21,84 +21,96 @@
...
 import java.util.Map;
 import java.util.concurrent.ExecutionException;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 @Slf4j
 @RuleNode(
         type = ComponentType.TRANSFORMATION,
-        name = "delete keys",
+        name = "delete key-values",
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java:43]:
```suggestion
        name = "delete key-value pairs",
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java [hunk 8]
```diff
@@ -21,84 +21,96 @@
...
-            ctx.tellSuccess(TbMsg.transformMsg(msg, metaData, msgData));
+                keysToDelete.forEach(metaDataMap::remove);
+                metaDataCopy = new TbMsgMetaData(metaDataMap);
+                break;
+            case DATA:
+                JsonNode dataNode = JacksonUtil.toJsonNode(msgData);
+                if (dataNode.isObject()) {
+                    ObjectNode msgDataObject = (ObjectNode) dataNode;
+                    dataNode.fields().forEachRemaining(entry -> {
+                        String keyData = entry.getKey();
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbDeleteKeysNode.java:91]:
```suggestion
                        String keyExpression = entry.getKey();
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbCopyKeysNodeTest.java [hunk 9]
```diff
@@ -156,6 +161,20 @@ void givenMsgDataNotJSONObject_whenOnMsg_thenTVerifyOutput() throws Exception {
...
+        var config = new TbCopyKeysNodeConfiguration().defaultConfiguration();
+        var oldConfigJson = (ObjectNode) JacksonUtil.valueToTree(config);
+        oldConfigJson.remove("copyFrom");
+        oldConfigJson.put("fromMetadata", "false");
+        // WHEN
+        TbPair<Boolean, JsonNode> upgrade = node.upgrade(0, oldConfigJson);
+        // THEN
+        assertTrue(upgrade.getFirst());
+        assertEquals(config, JacksonUtil.treeToValue(upgrade.getSecond(), config.getClass()));
+    }
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbCopyKeysNodeTest.java:176]: i think it is better to manually provide json string for both cases and set up parameterized test with two cases

Comment (dskarzh) [reply rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbCopyKeysNodeTest.java:176]: same in delete node test

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbRenameKeysNodeTest.java [hunk 10]
```diff
@@ -148,6 +153,20 @@ void givenMsgDataNotJSONObject_whenOnMsg_thenVerifyOutput() throws Exception {
...
+        var config = new TbRenameKeysNodeConfiguration().defaultConfiguration();
+        var oldConfigJson = (ObjectNode) JacksonUtil.valueToTree(config);
+        oldConfigJson.remove("renameIn");
+        oldConfigJson.put("fromMetadata", "false");
+        // WHEN
+        TbPair<Boolean, JsonNode> upgrade = node.upgrade(0, oldConfigJson);
+        // THEN
+        assertTrue(upgrade.getFirst());
+        assertEquals(config, JacksonUtil.treeToValue(upgrade.getSecond(), config.getClass()));
+    }
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/transform/TbRenameKeysNodeTest.java:168]: same comment as in other nodes

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java [hunk 11]
```diff
@@ -42,8 +40,15 @@
         name = "to email",
         configClazz = TbMsgToEmailNodeConfiguration.class,
         nodeDescription = "Transforms message to email message",
-        nodeDetails = "Transforms message to email message by populating email fields using values derived from message metadata. " +
-                      "Set 'SEND_EMAIL' output message type.",
+        nodeDetails = "Transforms message to email message. All email fields support templatization.<br><br>" +
+                "Additionally, supports adding inline base64 images to the mail body from message metadata " +
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java:44]:
```suggestion
                "Additionally, it is possible to add inline base64 images to the mail body from message metadata " +
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java [hunk 12]
```diff
@@ -42,8 +40,15 @@
...
+                "by extracting them from <strong><i>images</i></strong> metadata key-value.<br>" +
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java:45]:
```suggestion
                "by extracting them from <strong><i>images</i></strong> metadata key.<br>" +
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java [hunk 13]
```diff
@@ -42,8 +40,15 @@
...
+                "<strong><i>images</strong></i> key-value should represent a map of entries: <strong><i>cid --> encodedImage</i></strong>.<br>" +
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java:46]:
```suggestion
                "<strong><i>images</strong></i> value should represent a map of entries: <strong><i>cid --> encodedImage</i></strong>.<br>" +
```

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java [hunk 14]
```diff
@@ -42,8 +40,15 @@
...
+                "E.g., <strong><i>cid</i></strong> is a <strong><i>tb.example.png</i></strong> and<br>" +
+                "<strong><i>encodedImage</i></strong> is a <strong><i>data:image/png;base64,...</i></strong>, " +
+                "then you can use <i><strong>img</i></strong> HTML tag with src: <i><strong>cid:tb.example.png</i></strong>.<br><br>" +
+                "Set <code>SEND_EMAIL</code> output message type if transformation completed successfully.<br><br>"  +
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNode.java:50]:
```suggestion
                "Node sets <code>SEND_EMAIL</code> output message type if transformation completed successfully.<br><br>"  +
```

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNodeTest.java [hunk 15]
```diff
@@ -15,98 +15,151 @@
...
+        // WHEN
+        node.onMsg(ctxMock, msg);
+
+        // THEN
+        var msgCaptor = ArgumentCaptor.forClass(TbMsg.class);
+        var typeCaptor = ArgumentCaptor.forClass(TbMsgType.class);
+        var originatorCaptor = ArgumentCaptor.forClass(EntityId.class);
+        var metadataCaptor = ArgumentCaptor.forClass(TbMsgMetaData.class);
+        var dataCaptor = ArgumentCaptor.forClass(String.class);
+        verify(ctxMock, times(1)).transformMsg(
```
Comment (dskarzh) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/mail/TbMsgToEmailNodeTest.java:118]:
```suggestion
        verify(ctxMock).transformMsg(
```
image

verify is by default times 1

**Review by dskarzh — CHANGES_REQUESTED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbRenameKeysNode.java [hunk 16]
```diff
@@ -21,77 +21,94 @@
...
 @Slf4j
 @RuleNode(
         type = ComponentType.TRANSFORMATION,
         name = "rename keys",
+        version = 1,
         configClazz = TbRenameKeysNodeConfiguration.class,
-        nodeDescription = "Renames msg data or metadata keys to the new key names selected in the key mapping.",
-        nodeDetails = "If the key that is selected in the key mapping is missed in the selected msg source(data or metadata), it will be ignored." +
-                " Returns transformed messages via <code>Success</code> chain",
+        nodeDescription = "Renames message or message metadata key names.",
```
[the remark this answers is not in the document]
Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbRenameKeysNode.java:42]: TODO: check for all places where phrase "key names" added.

**Pull request merged**